agent/pkg/docker/docker.go

```diff
@@ -0,0 +1,464 @@
+	}
+
+	authString, err := registryToString(*auth)
+	if err != nil {
+		return fmt.Errorf("error encoding docker credentials: %w", err)
+	}
+
+	logs, err := d.cl.ImagePull(ctx, ref.String(), types.ImagePullOptions{
+		All:          false,
+		RegistryAuth: authString,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "error pulling image: %s", ref.String())
+	}
+	defer func() {
+		if err = logs.Close(); err != nil {
+			d.logger.WithError(err).Error("error closing log stream")
+		}
+	}()
+
+	if err = d.sendPullLogs(ctx, logs, p); err != nil {
+		return fmt.Errorf("error processing pull log stream: %w", err)
+	}
+	return nil
+}
+
+// CreateContainer creates a container according to the given spec, returning a docker container ID
+// to start it. It takes a caller-provided channel on which docker events are sent. Slow receivers
+// will block the call.
+func (d *Client) CreateContainer(
+	ctx context.Context,
+	req cproto.RunSpec,
+	p events.Publisher[Event],
+) (string, error) {
+	response, err := d.cl.ContainerCreate(
+		ctx, &req.ContainerConfig, &req.HostConfig, &req.NetworkingConfig, nil, "")
+	if err != nil {
+		return "", fmt.Errorf("creating container: %w", err)
+	}
+	id := response.ID
+	for _, w := range response.Warnings {
+		if err = p.Publish(ctx, NewLogEvent(model.LogLevelWarning, fmt.Sprintf(
+			"warning when creating container: %s", w,
+		))); err != nil {
+			return "", err
+		}
+	}
+
+	for _, copyArx := range req.Archives {
+		if err = p.Publish(ctx, NewLogEvent(model.LogLevelInfo, fmt.Sprintf(
+			"copying files to container: %s", copyArx.Path,
+		))); err != nil {
+			return "", err
+		}
+
+		files, cErr := archive.ToIOReader(copyArx.Archive)
+		if cErr != nil {
+			return "", fmt.Errorf("converting RunSpec Archive files to io.Reader: %w", cErr)
+		}
+		if err = d.cl.CopyToContainer(ctx, id, copyArx.Path, files, copyArx.CopyOptions); err != nil {
+			return "", fmt.Errorf("copying files to container: %w", err)
+		}
+	}
+	return id, nil
+}
+
+// RunContainer runs a container by docker container ID. It takes a caller-provided channel on which
+// docker events are sent. Slow receivers will block the call.
+func (d *Client) RunContainer(ctx context.Context, id string) (*Container, error) {
+	// Wait before start to not miss immediate exits.
+	waiter, errs := d.cl.ContainerWait(ctx, id, dcontainer.WaitConditionNextExit)
+
+	if err := d.cl.ContainerStart(ctx, id, types.ContainerStartOptions{}); err != nil {
+		return nil, fmt.Errorf("starting container: %w", err)
+	}
+
+	// If we specified a port to expose but not the host port to bind, Docker assigns an arbitrary host
+	// port, which we ask for here. (If we did specify a host port, this gives the same one back.)
+	containerInfo, err := d.cl.ContainerInspect(ctx, id)
+	if err != nil {
+		if cErr := d.RemoveContainer(ctx, id, true); cErr != nil {
+			d.logger.WithError(cErr).Errorf("removing container %s after inspect failure", id)
```

Comment:
can we also have the `id` as something like `WithField("dcl-container-id", id)`
(not sure about the field name, basically I don't want it to be confused with "our" container id)